fix #176601 unicode insert Upper & Lower Surrogates simultaneously

ericfont opened the pull request:
added new version of Text::insert which takes two chars as input, and inserts them together before incrementing the cursor, to prevent bug.

ericfont commented:
Note: I haven't actually testing this in mscore application...I will do that after I goto bed.  I've only tested this with my added mtests.  So you might want to hold off on accepting.  But I'm making PR now so you can take a look at how I changed the code.

Todo: is to figure out a way to test dropping a supplemental unicode char onto a text element...I have no idea how to do that, so if you have an idea, let me know.

lasconic commented:
If you mean how to call Text::drop then it's called when you press the alpha (F2) button and double click on one of the character.

ericfont commented:
ok.  Well in that case I'm trying out that Drop in musescore app and can still produce the error by dropping...I'm putting in the correction now and will repush.

I will have to study how Text::createLayout is used to determine how to have that support the supplementary unicode...but that might be a separate issue report.

ericfont commented:
ok, I got the drop behavior working, and that means that the bug descibed in the description for https://musescore.org/en/node/176601 is no longer present.  However, I still have to take care of pasting when inside musescore application as I described https://musescore.org/en/node/176601#comment-662016 which is not fixed yet...it seems that when copying a supplementary symbol to the unicode, that it doesn't put both chars of the symbol into the clipboard....

ericfont commented:
I've discovered that Text::selectedText() only returns one char of the pair when copying a suplementary unicode.

ericfont commented:
ok, fixed that specific reproduction steps for that copy paste bug.  But I can still produce a bug if I have two symbols: 𝄎𝄎, select all, Ctrl+C, move cursor in betweenen, and Ctrl+V, then produces ?'s.  That I and I find if I have 𝄎𝄎𝄎 and select only two of them, then the clipboard somehow picks up all three of them.  

So I'm getting closer, but we'll have to finalize this tomorrow.

ericfont commented:
I think I found a source of an error: TextBlock::insert is not inserting in correct spot.  It tries to insert into the text block at index of rcol, which is a cursor position.  But it needs to insert into spot at index of QChar position.

ericfont commented:
I'm going to have to have QList<TextFragment>::iterator TextBlock::fragment(int column, int* rcol) additionally return the actual QChar index relative to the TextFragment.  Currently it only returns rcol, the column index relative to the TexFragment.

ericfont commented:
going over this some more, I discovered a new bug present in 2.0.3: if inserting a Unicode after a FMUSL, then it won't add that as a new TextFragment of type TEXT, but will mistakenly append it to the last TextFragment which is a SYMBOL (and hence incompatible with TEXT) if the current state of the Cursor was in SYMBOL.  That means the Unicode is not displayed, bucause it gets appened to the end of a text frament.  

See the end of void TextBlock::insert(TextCursor* cursor, const QString& s) ...and notice the if block:

            if (!_text.empty() && _text.back().format == *cursor->format())
                  _text.back().text.append(s);
            else
                  _text.append(TextFragment(cursor, s));

That I believe is incorrect, because I think it should test for not whether the cursor is of the same type, but should test whether the TextFragment's type is TEXT or SYMBOL:

            if (!_text.empty() && _text.back().format.type() == CharFormatType::TEXT)
                  _text.back().text.append(s);
            else
                  _text.append(TextFragment(cursor, s));

Minimum reproduction steps are:
-new staff text
-press F2 and drop one or more FMUSL symbols
-click outside of text box
-click inside text box and position cursor at end
-press F2 and drop two Supplementary Unicode symbols

result: not displayed...

I'm going to piggyback a fix for this with this PR while I'm looking at it...

ericfont commented:
I'm still trying to figure why the previous code for both TextBlock::insert methods sometimes tests for cursor's CharFormatType and why sometimes tests the current TextFragment's CharFormatType.  I would think the two versions of TextBlock::insert would be mirror images of eachother.

ericfont commented:
I believe I fixed that bug I just described.  Fixed by making sure to append a new TextFragment regadless of state of cursor, and also making TextFragment::TextFragment(TextCursor* cursor, const QString& s) constructor set the type to CharFormatType::TEXT.  Now I need to make sure I didn't break something...I will need to add mtests now...

ericfont commented:
But I think i broke something...I notice title text is not the same font style as it should be...

ericfont commented:
just a note: I testing inputting simplified Chinese characters on windows, and looked in deubgger and saw that they are not using a Supplementary Plane:
						[0]	'我' 	25105	0x6211	QChar
						[1]	'爱' 	29233	0x7231	QChar
						[2]	'M' 	77	0x004d	QChar
						[3]	'u' 	117	0x0075	QChar
						[4]	's' 	115	0x0073	QChar
						[5]	'e' 	101	0x0065	QChar
						[6]	's' 	115	0x0073	QChar
						[7]	'c' 	99	0x0063	QChar
						[8]	'o' 	111	0x006f	QChar
						[9]	'r' 	114	0x0072	QChar
						[10]	'e' 	101	0x0065	QChar
						[11]	'你' 	20320	0x4f60	QChar
						[12]	'好' 	22909	0x597d	QChar
						[13]	'。' 	12290	0x3002	QChar
And wikipedia says there are 20,000+ characters that all fit in the Basic Plane  https://en.wikipedia.org/wiki/CJK_Unified_Ideographs which are a common subset.  So that probably explains why haven't run into a problem with CJK that aren't in a Supplementary Plane.

ericfont commented:
@lasconic I believe this is done.  I haven't been able to produce any more errors manipulating Supplemental unicode in mscore app.  Here is the commit message:

Any operations dealing with Supplemental Unicode must operate on high & low surrogates simulatenously, since they belong together as a pair:
-Text::insert method has a new version of which takes two QChars as input (the high & low surrogates), and inserts them together before incrementing the cursor.
-TextBlock::fragment method now has return parameter ridx (relative QChar index) in addition to rcol (relative cursor position).  For Basic Unicode TextFragments, ridx will be same as rcol, since each Basic Unicode char is one QChar.  But for text that has Supplemental Unicode, ridx will be be different since each Supplemental Unicode char is two QChars.
-TextBlock::text method now includes any high surrogate QChars in the returned string.  Previously, this method simply skipped over QChars that were high surrogates.  When iterating over the TextFragment, only increments col when comming across a QChar that is not a high surrogate.
-Text::createLayout, Text::paste and Text::drop methods now uses the new Text::insert method to insert both high & low surrogates together.  Previously, the cursor position would get messed up when each QChar was added seperately.
-TextBlock::remove(int start, int n) will now add both QChars of the surrogate pair to the return string.

Also was issue with adding text after a SMUFL symbol if the cursor mode was already in SMUFL.  Fixed by making sure to append a new TextFragment regardless of state of cursor, and also making TextFragment::TextFragment(TextCursor* cursor, const QString& s) constructor set the type to CharFormatType::TEXT.

ericfont commented on mtest/libmscore/text/tst_text.cpp:
```diff
@@ -497,8 +504,243 @@ void TestText::testMixedTypesDeletePreviousChar()
+      void setLine(int val)         { _line = val; }
+      void setColumn(int val)       { _column = val; }
+      void setSelectLine(int val)   { _selectLine = val; }
+      void setSelectColumn(int val) { _selectColumn = val; }*/
```
I forgot to delete the commented lines, will do before final push.

ericfont replied:
I guess the next modifier of tst_text.cpp will delete these commented lines.

lasconic replied:
Yes, I will
